cl_kernel/kernel_3d_denoise_slm.cl

```diff
@@ -0,0 +1,215 @@
+/*
+ * function: kernel_3d_denoise_slm
+ *     3D Noise Reduction
+ * gain:        The parameter determines the filtering strength for the reference block
+ * threshold:   Noise variances of observed image
+ * restoredPrev: The previous restored image, image2d_t as read only
+ * output:      restored image, image2d_t as write only
+ * input:       observed image, image2d_t as read only
+ * inputPrev1:  reference image, image2d_t as read only
+ * inputPrev2:  reference image, image2d_t as read only
+ */
+
+#define WORKGROUP_WIDTH    8
+#define WORKGROUP_HEIGHT   8
+
+#define REF_BLOCK_X_OFFSET  1
+#define REF_BLOCK_Y_OFFSET  4
+
+#define REF_BLOCK_WIDTH  (WORKGROUP_WIDTH + 2 * REF_BLOCK_X_OFFSET)
+#define REF_BLOCK_HEIGHT (WORKGROUP_HEIGHT + 2 * REF_BLOCK_Y_OFFSET)
+
+
+inline void weighted_average (__read_only image2d_t input,
+                              bool load_observe,
+                              float4* restore,
+                              float2* sum_weight,
+                              float gain,
+                              float threshold)
+{
+    sampler_t sampler = CLK_NORMALIZED_COORDS_FALSE | CLK_ADDRESS_CLAMP_TO_EDGE | CLK_FILTER_NEAREST;
+
+    const int local_id_x = get_local_id(0);
+    const int local_id_y = get_local_id(1);
+    const int group_id_x = get_group_id(0);
+    const int group_id_y = get_group_id(1);
+
+    float4 cur_block = 0.0f;
+    __local float4 local_ref_cache[REF_BLOCK_HEIGHT * REF_BLOCK_WIDTH];
+    __local float4 local_obs_cache[WORKGROUP_HEIGHT * WORKGROUP_WIDTH];
+
+    int start_x = mad24(group_id_x, WORKGROUP_WIDTH, -REF_BLOCK_X_OFFSET);
+    int start_y = mad24(group_id_y, WORKGROUP_HEIGHT, -REF_BLOCK_Y_OFFSET);
+
+    for (int i = 0; i < REF_BLOCK_HEIGHT; i++) {
+        for (int j = local_id_x; j < REF_BLOCK_WIDTH; j += WORKGROUP_WIDTH) {
+            int corrd_x = start_x + (j % REF_BLOCK_WIDTH);
+            local_ref_cache[i * REF_BLOCK_WIDTH + j] = read_imagef(input, sampler, (int2)(corrd_x, start_y));
+        }
+        start_y += 1;
+    }
```

Comment:
double for-loop slow down performance and cause same slm-cache frequently re-write/re-read very bad. please fix it first.
In this case there are 8x8=64 workitems in a workgroup. and would load 10x16=160 pixels. the best load only need each workitem read (160/64 = )3 times. but this double-for-loop need each workitem load 10/2 x 10 = 50times which slow down the performance heavily.